### PR TITLE
Story198: Grid Measurement Tool

### PR DESCRIPTION
**Pull request by mj0stjo**
# Story198: Implementing the grid reference tool
## Workflow
- Select the grid reference tool
- grid and some controls are displayed
- rotate the grid until it fits into your target plane
- click on "ebene setzen"
- select a reference distance by selecting two points and specifying the length (just like in the rectangle reference)
- the reference is set and you can measure
## Limitations
- Basic Undo and Redo functionalities work but I did not test every edge case so I expect it to not work perfectly
- There is no line preview for selecting the reference distance as I do not know how it works

### CODE REVIEW

**Comment by Markt301** [comment]
I have to zoom out in order to display all the controls:
grafik

**Comment by Markt301** [comment]
I have to zoom out in order to display all the controls:
grafik

**Comment by Markt301** [comment]
For the Preview:
- If there is time left I can do the Preview for this Tool

**Comment by Markt301** [comment]
It would be nice if you were nofified that you can now start with setting the reference distance. Maybe a Toast?

**Comment by mj0stjo** [comment]
> For the Preview:
> 
>     * If there is time left I can do the Preview for this Tool

I created an issue (#133) for this. If there is spare time in the end we (or better said you), can implement it, otherwise I think it's still fine....

**Comment by mj0stjo** [comment]
> * If you set the 'ebene' and make a point or two
> 
>     * Then go back to the beginning (`Crl + Z`)
> 
>     * And go forth twice (`Crl + Y`)
> 
>     * You will set a point
> 
> 
> But this bug is not that important

Yeah, I think this bug is not that sraight forward to fix, even though I know why it is like this...
I will create a new issue but I don't think this will be fixed anymore.

**Comment by mj0stjo** [comment]
> It would be nice if you were nofified that you can now start with setting the reference distance. Maybe a Toast?

very good idea, I added a toast!

**Comment by mj0stjo** [comment]
> I have to zoom out in order to display all the controls

I tried to resolve this by reducing the knob sizes, unfortunately these vue components don't scale very well dynamically...

**Comment by Markt301** [comment]
just for the protocol:
- if you set the 'ebene' and then select the rectangle Reference Tool the Länge-Auswählen-Popup will appear:
grafik

**Comment by Markt301** [comment]
if you:
- set the 'ebene'
- Dont draw anything
- Then select the Grid Reference again
- And also set this 'ebene'
- You can only make one Point and are asked the put in the distance

File: WorkerView/src/components/Editor/Logic/Tools/GridReferenceTool.js [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+        src.push([point2.x, point2.y]);
+        src.push([point3.x, point3.y]);
+        src.push([point4.x, point4.y]);
+
+        // Generating the "destination" array which is the coordinates of the reference in the real world which match to the source points.
+        // These coordinates can be calculated based on the edge lengths specified by the user.
+        var dst = [[0, 100], [100, 100], [100, 0], [0, 0]];
+
+        // Entzerrt aber ohne skalierung:
+        var entzerrt_ohne_scale = calculatePerspectiveMatrix(src, dst)
```
Comment (Markt301) [inline WorkerView/src/components/Editor/Logic/Tools/GridReferenceTool.js:116]: interesting name :D

File: WorkerView/src/components/Editor/MainEditor.vue [hunk 2]
```diff
@@ -239,8 +249,44 @@ function canvasBack(event){
     controller.redo()
   }
 
+  else if (event.key === 'f' || event.keyCode === 'F') {
```
Comment (Markt301) [inline WorkerView/src/components/Editor/MainEditor.vue:252]: is this needed?

Comment (mj0stjo) [reply WorkerView/src/components/Editor/MainEditor.vue]: Nope, I removed it!

**Review by Markt301 — APPROVED** [review]
I have to say:
- Nice Job!
- Approved!
